packages/@lwc/template-compiler/src/codegen/codegen.ts

```diff
@@ -536,25 +539,69 @@ export default class CodeGen {
 
         const args: t.Expression[] = [t.callExpression(identifier, []), t.literal(key)];
 
-        // Only add the third argument (databag) if this element needs it
-        if (element.listeners.length || element.directives.length) {
-            const databagProperties: t.Property[] = [];
+        // Only add the third argument (staticParts) if this element needs it
+        const staticParts = this.genStaticParts(element);
+        if (staticParts) {
+            args.push(staticParts);
+        }
+
+        return this._renderApiCall(RENDER_APIS.staticFragment, args);
+    }
 
-            // has event listeners
-            if (element.listeners.length) {
-                databagProperties.push(this.genEventListeners(element.listeners));
+    genStaticParts(element: StaticElement): t.ArrayExpression | undefined {
+        const stack: (StaticElement | Text)[] = [element];
+        const partIdsToDatabagProps = new Map<number, t.Property[]>();
+        let partId = -1;
+
+        const addDatabagProp = (prop: t.Property) => {
+            let databags = partIdsToDatabagProps.get(partId);
+            if (!databags) {
+                databags = [];
+                partIdsToDatabagProps.set(partId, databags);
             }
+            databags.push(prop);
+        };
+
+        // Depth-first traversal. We assign a partId to each element, which is an integer based on traversal order.
+        while (stack.length > 0) {
+            const node = stack.shift()!;
+            partId++;
+            if (isElement(node)) {
+                // has event listeners
+                if (node.listeners.length) {
+                    addDatabagProp(this.genEventListeners(node.listeners));
+                }
 
-            // see STATIC_SAFE_DIRECTIVES for what's allowed here
-            for (const directive of element.directives) {
-                if (directive.name === 'Ref') {
-                    databagProperties.push(this.genRef(directive));
+                // see STATIC_SAFE_DIRECTIVES for what's allowed here
+                for (const directive of node.directives) {
+                    if (directive.name === 'Ref') {
```

Comment:
I don't think so, no. Because even if we support more `STATIC_SAFE_DIRECTIVES` in the future, we would need to do something different for each one of them. In this case, it's `genRef` that we have to do.

Reply:
Got it thanks!